android/examples/unrooted_android.c

```diff
@@ -0,0 +1,300 @@
+{
+    struct libusb_device_descriptor desc;
+    unsigned char string[256];
+    const char *speed;
+    int ret;
+    uint8_t i;
+
+    switch (libusb_get_device_speed(dev)) {
+        case LIBUSB_SPEED_LOW:		speed = "1.5M"; break;
+        case LIBUSB_SPEED_FULL:		speed = "12M"; break;
+        case LIBUSB_SPEED_HIGH:		speed = "480M"; break;
+        case LIBUSB_SPEED_SUPER:	speed = "5G"; break;
+        case LIBUSB_SPEED_SUPER_PLUS:	speed = "10G"; break;
+        default:			speed = "Unknown";
+    }
+
+    ret = libusb_get_device_descriptor(dev, &desc);
+    if (ret < 0) {
+        LOGD("failed to get device descriptor");
+        return;
+    }
+
+    LOGD("Dev (bus %u, device %u): %04X - %04X speed: %s\n",
+           libusb_get_bus_number(dev), libusb_get_device_address(dev),
+           desc.idVendor, desc.idProduct, speed);
+
+    if (!handle)
+        libusb_open(dev, &handle);
+
+    if (handle) {
+        if (desc.iManufacturer) {
+            ret = libusb_get_string_descriptor_ascii(handle, desc.iManufacturer, string, sizeof(string));
+            if (ret > 0)
+                LOGD("  Manufacturer:              %s\n", (char *)string);
+        }
+
+        if (desc.iProduct) {
+            ret = libusb_get_string_descriptor_ascii(handle, desc.iProduct, string, sizeof(string));
+            if (ret > 0)
+                LOGD("  Product:                   %s\n", (char *)string);
+        }
+
+        if (desc.iSerialNumber && verbose) {
+            ret = libusb_get_string_descriptor_ascii(handle, desc.iSerialNumber, string, sizeof(string));
+            if (ret > 0)
+                LOGD("  Serial Number:             %s\n", (char *)string);
+        }
+    }
+
+    if (verbose) {
+        for (i = 0; i < desc.bNumConfigurations; i++) {
+            struct libusb_config_descriptor *config;
+
+            ret = libusb_get_config_descriptor(dev, i, &config);
+            if (LIBUSB_SUCCESS != ret) {
+                LOGD("  Couldn't retrieve descriptors\n");
+                continue;
+            }
+
+            print_configuration(config);
+
+            libusb_free_config_descriptor(config);
+        }
+
+        if (handle && desc.bcdUSB >= 0x0201)
+            print_bos(handle);
+    }
+
+    if (handle)
+        libusb_close(handle);
+}
+
+
+// fileDescriptor = is the native File Descriptor obtained in Java and transfered to native over JNA for Example.
+int unrooted_usb_description(int fileDescriptor)
+{
+    libusb_context *ctx;
+    libusb_device_handle *devh = NULL;
+    int r = 0;
+    verbose = 1;
+    r = libusb_set_option(&ctx, LIBUSB_OPTION_WEAK_AUTHORITY, NULL);
+    if (r != LIBUSB_SUCCESS) {
+        LOGD("libusb_init failed: %d\n", r);
+        return -1;
+    }
+    r = libusb_init(&ctx);
+    if (r < 0) {
+        LOGD("libusb_init failed: %d\n", r);
+        return r;
+    }
+    r = libusb_wrap_sys_device(NULL, (intptr_t)fileDescriptor, &devh);
```

Comment:
Using NULL in place of the context crashes my example based on unrooted_android.c of this PR.

This line is fine...
`r = libusb_wrap_sys_device(ctx, (intptr_t)fileDescriptor, &devh);`


Code used :
```
int unrooted_usb_description(int fileDescriptor) {
    libusb_context *ctx = NULL;
    libusb_device_handle *devh = NULL;
    int r = 0;
    LOGD("Hello here !!");
    r = libusb_set_option(ctx, LIBUSB_OPTION_WEAK_AUTHORITY, NULL);
    if (r != LIBUSB_SUCCESS) {
        LOGD("libusb_set_option failed: %d\n", r);
        return -1;
    }
    LOGD("libusb_set_option OK: %d\n", r);
    r = libusb_init(&ctx);
    if (r < 0) {
        LOGD("libusb_init failed: %d\n", r);
        return r;
    }
    LOGD("libusb_init OK: %d\n", r);
    r = libusb_wrap_sys_device(ctx, (intptr_t)fileDescriptor, &devh);
    if (r < 0) {
        LOGD("libusb_wrap_sys_device failed: %d\n", r);
        return r;
    } else if (devh == NULL) {
        LOGD("libusb_wrap_sys_device returned invalid handle\n");
        return r;
    }
    LOGD("libusb_wrap_sys_device OK: %d\n", r);
    ...
```

Here is the logcat:
```
27646 27646 I Libusb_Sample: USB test permission: "/dev/bus/usb/001/002": true
27646 27646 I Libusb_Sample: USB file descriptor: "/dev/bus/usb/001/002": 57
27646 27646 D LibUsb  : Hello here !!
27646 27646 D LibUsb  : libusb_set_option OK: 0
27646 27646 D LibUsb  : libusb_init OK: 0
27646 27646 W MY.APP: type=1400 audit(0.0:9118): avc: denied { read } for name="usb" dev="tmpfs" ino=122799 scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:object_r:usb_device:s0 tclass=dir permissive=0
27646 27646 W MY.APP: type=1400 audit(0.0:9119): avc: denied { read } for name="/" dev="tmpfs" ino=13042 scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:object_r:device:s0 tclass=dir permissive=0
27646 27646 F libc    : Fatal signal 11 (SIGSEGV), code 1, fault addr 0x1a8 in tid 27646 (MY.APP)
  441   441 W         : debuggerd: handling request: pid=27646 uid=10143 gid=10143 tid=27646
27905 27905 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
27905 27905 F DEBUG   : Build fingerprint: 'Xiaomi/santoni/santoni:7.1.2/N2G47H/V11.0.2.0.NAMMIXM:user/release-keys'
27905 27905 F DEBUG   : Revision: '0'
27905 27905 F DEBUG   : ABI: 'arm64'
27905 27905 F DEBUG   : pid: 27646, tid: 27646, name: MY.APP  >>> com.MY.APP <<<
27905 27905 F DEBUG   : signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x1a8
27905 27905 F DEBUG   :     x0   00000000000001a8  x1   0000000000000004  x2   0000007f80166800  x3   0000007f801677d4
27905 27905 F DEBUG   :     x4   0000007ff1fb2760  x5   0000000000000004  x6   0000007f6a0a09ca  x7   10c4400000000110
27905 27905 F DEBUG   :     x8   0000000000000000  x9   0000007f6a112780  x10  69440961db40bea4  x11  00000000fffffff4
27905 27905 F DEBUG   :     x12  00000000fffffff4  x13  0000007f80167340  x14  0000000000000041  x15  000000000000004c
27905 27905 F DEBUG   :     x16  0000007f80208038  x17  0000007f8b150b3c  x18  0000000000000000  x19  0000007f6a8dce38
27905 27905 F DEBUG   :     x20  0000007f6a8dcd30  x21  0000007ff1fb3568  x22  0000000000000000  x23  0000000000000008
27905 27905 F DEBUG   :     x24  0000007ff1fb2d38  x25  0000000000000001  x26  000000000000001e  x27  0000007ff1fb2ad0
27905 27905 F DEBUG   :     x28  0000007ff1fb2a10  x29  0000007ff1fb27c0  x30  0000007f80191aa4
27905 27905 F DEBUG   :     sp   0000007ff1fb27b0  pc   0000007f8b150b3c  pstate 0000000060000000
27905 27905 F DEBUG   : 
27905 27905 F DEBUG   : backtrace:
27905 27905 F DEBUG   :     #00 pc 0000000000068b3c  /system/lib64/libc.so (pthread_mutex_lock)
27905 27905 F DEBUG   :     #01 pc 0000000000071aa0  /data/app/com.MY.APP-2/lib/arm64/libnative_wrapper.so
27905 27905 F DEBUG   :     #02 pc 000000000007134c  /data/app/com.MY.APP-2/lib/arm64/libnative_wrapper.so
27905 27905 F DEBUG   :     #03 pc 00000000000798c8  /data/app/com.MY.APP-2/lib/arm64/libnative_wrapper.so
27905 27905 F DEBUG   :     #04 pc 000000000007648c  /data/app/com.MY.APP-2/lib/arm64/libnative_wrapper.so
27905 27905 F DEBUG   :     #05 pc 000000000006b5c8  /data/app/com.MY.APP-2/lib/arm64/libnative_wrapper.so (libusb_wrap_sys_device+184)
27905 27905 F DEBUG   :     #06 pc 00000000000676a8  /data/app/com.MY.APP-2/lib/arm64/libnative_wrapper.so (unrooted_usb_description+320)
27905 27905 F DEBUG   :     #07 pc 0000000000010154  /data/app/com.MY.APP-2/lib/arm64/libjnidispatch.so
27905 27905 F DEBUG   :     #08 pc 000000000000f938  /data/app/com.MY.APP-2/lib/arm64/libjnidispatch.so
27905 27905 F DEBUG   :     #09 pc 0000000000005cfc  /data/app/com.MY.APP-2/lib/arm64/libjnidispatch.so
27905 27905 F DEBUG   :     #10 pc 0000000000007b68  /data/app/com.MY.APP-2/lib/arm64/libjnidispatch.so (Java_com_sun_jna_Native_invokeInt+32)
27905 27905 F DEBUG   :     #11 pc 000000000056db04  /data/app/com.MY.APP-2/oat/arm64/base.odex (offset 0x512000)
```

Note that the 2 lines with `avc: denied` (SE Linux ?) are also present when It's working with the ctx given to the function.

Maybe this will help also for #874

Reply:
Ok, passing the context in this function makes sense.

About #874 I think this PR will work without libusb_wrap_sys_device.
There are modified versions of libusb, which uses other approaches to get the library working on Android:
https://github.com/saki4510t/UVCCamera